File: android/app/src/test/java/ai/oriveo/community/feature/chat/components/UserMessageFoldTest.kt
package ai.oriveo.community.feature.chat.components

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/** The long user message fold rule; the web client's `user-message-fold.test.ts` checks the same cases. */
class UserMessageFoldTest {

    private fun arabic(length: Int): String {
        val sentence = "هذا نص عربي طويل جدا بدون أي سطر جديد. "
        return buildString { while (this.length < length) append(sentence) }
    }

    @Test
    fun `threshold boundary counts UTF-16 units`() {
        val atThreshold = "a".repeat(UserMessageFold.THRESHOLD_UTF16)
        assertFalse(UserMessageFold.shouldFold(atThreshold))
        assertTrue(UserMessageFold.shouldFold(atThreshold + "a"))
        assertFalse(UserMessageFold.shouldFold("😀".repeat(3_000)))
        assertTrue(UserMessageFold.shouldFold("😀".repeat(3_001)))
    }

    @Test
    fun `preview is a bounded prefix of the original`() {
        val text = arabic(200_000)
        val preview = UserMessageFold.preview(text)
        assertTrue(text.startsWith(preview))
        assertTrue(preview.length <= UserMessageFold.PREVIEW_UTF16)
        assertTrue(preview.length > UserMessageFold.PREVIEW_UTF16 - 30)
    }

    @Test
    fun `preview never splits an emoji sequence or a base letter from its marks`() {
        val family = "👩‍👩‍👧‍👦"
        val text = "a".repeat(UserMessageFold.PREVIEW_UTF16 - 1) + family + "b".repeat(8_000)
        assertEquals("a".repeat(UserMessageFold.PREVIEW_UTF16 - 1), UserMessageFold.preview(text))

        // ب + kasra + shadda is one grapheme; the prefix may only stop between graphemes.
        val marked = "بِّ".repeat(3_000)
        assertEquals(0, UserMessageFold.preview(marked).length % 3)
    }

    @Test
    fun `preview keeps at most 60 line breaks`() {
        val text = "line\n".repeat(5_000)
        val preview = UserMessageFold.preview(text)
        assertEquals(UserMessageFold.PREVIEW_LINE_CAP, preview.count { it == '\n' })
        assertTrue(text.startsWith(preview))
    }

    @Test
    fun `reading chunks keep every character and stay under the chunk cap`() {
        val text = arabic(200_000) + "\n\nsecond paragraph\n" + "한".repeat(5_000)
        val chunks = UserMessageFold.readingChunks(text)
        assertEquals(text.replace("\n", ""), chunks.joinToString(""))
        assertTrue(chunks.all { it.length <= UserMessageFold.READING_CHUNK_UTF16 })
        // A 200,000-character paragraph becomes about 100 chunks; the LazyColumn composes only the visible few.
        assertTrue(chunks.size in 100..140)
        // The Arabic text has periods and spaces: breaks land after a sentence end, so no chunk starts with whitespace.
        assertTrue(chunks.first().endsWith(". "))
    }

    @Test
    fun `reading chunks break long runs without spaces on grapheme boundaries`() {
        val text = "👍🏽".repeat(3_000)
        val chunks = UserMessageFold.readingChunks(text)
        assertEquals(text, chunks.joinToString(""))
        chunks.forEach { chunk -> assertEquals(0, chunk.length % 4) }
    }
}

File: web/apps/app/lib/core/chat/user-message-fold.test.ts
import { describe, expect, it } from 'vitest';
import {
  USER_MESSAGE_FOLD_THRESHOLD,
  USER_MESSAGE_PREVIEW_LENGTH,
  USER_MESSAGE_PREVIEW_LINE_CAP,
  USER_MESSAGE_READING_CHUNK_LENGTH,
  shouldFoldUserMessage,
  userMessagePreview,
  userMessageReadingChunks,
} from './user-message-fold';

// Mirrors the fold tests of the native clients.
const arabic = (length: number) => {
  const sentence = 'هذا نص عربي طويل جدا بدون أي سطر جديد. ';
  let out = '';
  while (out.length < length) out += sentence;
  return out;
};

describe('user message fold', () => {
  it('folds only past the UTF-16 threshold', () => {
    const atThreshold = 'a'.repeat(USER_MESSAGE_FOLD_THRESHOLD);
    expect(shouldFoldUserMessage(atThreshold)).toBe(false);
    expect(shouldFoldUserMessage(`${atThreshold}a`)).toBe(true);
    expect(shouldFoldUserMessage('😀'.repeat(3_000))).toBe(false);
    expect(shouldFoldUserMessage('😀'.repeat(3_001))).toBe(true);
  });

  it('keeps a bounded prefix of the original', () => {
    const text = arabic(200_000);
    const preview = userMessagePreview(text);
    expect(text.startsWith(preview)).toBe(true);
    expect(preview.length).toBeLessThanOrEqual(USER_MESSAGE_PREVIEW_LENGTH);
    expect(preview.length).toBeGreaterThan(USER_MESSAGE_PREVIEW_LENGTH - 30);
  });

  it('never splits an emoji sequence or a base letter from its marks', () => {
    const family = '👩‍👩‍👧‍👦';
    const text = `${'a'.repeat(USER_MESSAGE_PREVIEW_LENGTH - 1)}${family}${'b'.repeat(8_000)}`;
    expect(userMessagePreview(text)).toBe('a'.repeat(USER_MESSAGE_PREVIEW_LENGTH - 1));
    // ب + kasra + shadda is one grapheme.
    expect(userMessagePreview('بِّ'.repeat(3_000)).length % 3).toBe(0);
  });

  it('keeps at most 60 line breaks and counts CRLF once', () => {
    const lf = userMessagePreview('line\n'.repeat(5_000));
    expect(lf.split('\n').length - 1).toBe(USER_MESSAGE_PREVIEW_LINE_CAP);
    const crlf = userMessagePreview('line\r\n'.repeat(5_000));
    expect(crlf.split('\r\n').length - 1).toBe(USER_MESSAGE_PREVIEW_LINE_CAP);
  });

  it('splits the full text into bounded reading chunks without losing characters', () => {
    const text = `${arabic(200_000)}\n\nsecond paragraph\n${'한'.repeat(5_000)}`;
    const chunks = userMessageReadingChunks(text);
    expect(chunks.join('')).toBe(text.replaceAll('\n', ''));
    expect(chunks.every((chunk) => chunk.length <= USER_MESSAGE_READING_CHUNK_LENGTH)).toBe(true);
    expect(chunks.length).toBeGreaterThanOrEqual(100);
    expect(chunks[0].endsWith('. ')).toBe(true);

    const emoji = '👍🏽'.repeat(3_000);
    const emojiChunks = userMessageReadingChunks(emoji);
    expect(emojiChunks.join('')).toBe(emoji);
    expect(emojiChunks.every((chunk) => chunk.length % 4 === 0)).toBe(true);
  });
});
